### PR TITLE
chore: backport documentation changes to v3.5.2

### PR DESCRIPTION
**Pull request by jmeridth**
### Motivation

relates to #12414

### Modifications

- [x] move over files to help with readthedocs build
  - [x] .readthedocs.yaml
  - [x] Makefile
  - [x] mkdocs.yml
  - [x] hack/check-mkdocs.sh
  - [x] docs/requirements.txt
- [x] change links from github pages to readthedocs v3.5.2

### CODE REVIEW

**Review by terrytangyuan — APPROVED** [review]

**Comment by terrytangyuan** [comment]
LGTM

**Comment by jmeridth** [comment]
@terrytangyuan will we move the tag for v3.5.2 to this new SHA (aka HEAD of release-3.5.2 branch)?

**Comment by terrytangyuan** [comment]
I don't think that's possible

**Comment by terrytangyuan** [comment]
We should only create new tags instead of modifying existing ones

**Comment by agilgur5** [comment]
Ah looks like Argo CD has the `release-3.5` -style branch for each minor that I think it would make sense for us to use as well. See my comment in https://github.com/argoproj/argo-workflows/issues/12414#issuecomment-1876216461 for screenshots/examples of that

File: README.md [hunk 1]
```diff
@@ -20,11 +20,11 @@ Argo is a [Cloud Native Computing Foundation (CNCF)](https://cncf.io/) graduated
 
 ## Use Cases
 
-* [Machine Learning pipelines](https://argoproj.github.io/argo-workflows/use-cases/machine-learning/)
-* [Data and batch processing](https://argoproj.github.io/argo-workflows/use-cases/data-processing/)
-* [Infrastructure automation](https://argoproj.github.io/argo-workflows/use-cases/infrastructure-automation/)
-* [CI/CD](https://argoproj.github.io/argo-workflows/use-cases/ci-cd/)
-* [Other use cases](https://argoproj.github.io/argo-workflows/use-cases/other/)
+* [Machine Learning pipelines](https://argo-workflows.readthedocs.io/en/v3.5.2/use-cases/machine-learning/)
```
Comment (agilgur5) [inline README.md:23]: Ah I see, these changes will impact people who navigate to the `release-3.5` branch directly for the examples and docs.

The source code changes would only affect the next version of the binary though, which would be v3.5.3 at that time. Hmmm... using `release-3.5` or `3.5.x` for the version number in the docs permalink may make the most sense then

**Comment by terrytangyuan** [comment]
We do have those release branches. The problem is modifying an existing tag.

**Comment by terrytangyuan** [comment]
See https://github.com/argoproj/argo-workflows/branches/all?query=release

**Comment by agilgur5** [comment]
> See https://github.com/argoproj/argo-workflows/branches/all?query=release

It looks like we're currently using `release-3.5` as tracking v3.5.0 specifically though (the `.0` patch as in), as opposed to the minor as a whole, since it is behind `release-3.5.2`

**Comment by terrytangyuan** [comment]
Can we use a branch? Then it is as simple as syncing release-3.5 branch to be up-to-date with 3.5.2

**Comment by agilgur5** [comment]
Looks like we can indeed use a branch:

Screenshot 2024-01-03 at 9 50 07 PM

**Comment by jmeridth** [comment]
Yes we can but I need to issue another PR to change the URL to `/release-3.5.2` instead of `/v3.5.2`. Why can't we move the tag? I'm not understanding that. Shouldn't the tag point to the head of its release branch?

**Comment by jmeridth** [comment]
Nm. I see @agilgur5's comment [here](https://github.com/argoproj/argo-workflows/issues/12414#issuecomment-1876216461). We should match ArgoCD

**Comment by agilgur5** [comment]
> Why can't we move the tag? I'm not understanding that.

A tag is _supposed_ to be a stable reference, as opposed to a branch. Some artifact registries don't allow changing tags either (like NPM, for instance).
Git/Go and Docker/OCI registries actually have _mutable_ references however, which is why using a SHA or digest in addition to the version number is a general recommendation for supply chain security purposes for those registries.

But for _most_ intents, you typically want tags to be **immutable**.
This is an edge case where I'd support moving it, although as we're not making a new GH release of the binary etc, I'd rather keep things consistent where possible. And if we're just doing the minor branch then we can leave the tag as immutable.

The branch won't cover all the bases (i.e. old versions of the binary and old release tags will still point to GH Pages), but it'll cover all the most important ones, which is "good enough" IMO. Everything else will get redirected to `latest`, which matches the previous behavior anyway

**Comment by agilgur5** [comment]
Follow-up PR in #12460

**Pull request merged**